File: 5a.py
#! /usr/bin/python3
import sys

grid = {}
danger_tally = 0

def inc(arr, x, y):
    if not x in arr:
        arr[x] = {}
    if not y in arr[x]:
        grid[x][y] = 0
    grid[x][y] += 1
    return 1 if grid[x][y] == 2 else 0
    
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        ((x1, y1), (x2, y2)) = map(lambda coord: map(int, coord.split(',')),
                                   line.split(' -> '))
        if x1 != x2 and y1 != y2:
            continue
        if x1 == x2 :
            if y2 < y1:
                 (y1, y2) = (y2, y1)
            for y in range(y1, y2 + 1):
                danger_tally += inc(grid, x1, y)
        elif y1 == y2:
            if x2 < x1:
                (x1, x2) = (x2, x1)
            for x in range(x1, x2 + 1):
                danger_tally += inc(grid, x, y1)

print(danger_tally)

File: 5a.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my @grid;
my $danger_tally = 0;
while (<>) {
    my ($x1, $y1, $x2, $y2) = /(.+),(.+) -> (.+),(.+)/;
    next unless $x1 == $x2 || $y1 == $y2;

    if ($x1 == $x2) {
        ($y1, $y2) = ($y2, $y1) if $y2 < $y1;
        for my $y ($y1 .. $y2) {
            ++$danger_tally if 1 == $grid[$x1][$y]++;
        }
    } elsif ($y1 == $y2) {
        ($x1, $x2) = ($x2, $x1) if $x2 < $x1;
        for my $x ($x1 .. $x2) {
            ++$danger_tally if 1 == $grid[$x][$y1]++;
        }
    }
}
say $danger_tally;

__DATA__
0,9 -> 5,9
8,0 -> 0,8
9,4 -> 3,4
2,2 -> 2,1
7,0 -> 7,4
6,4 -> 2,0
0,9 -> 2,9
3,4 -> 1,4
0,0 -> 8,8
5,5 -> 8,2
